hicommonwealth/commonwealth
# Auto-upgrade Community to v1 if needed

## Link to Issue
Closes: #6070 
Closes: #3292

## Description of Changes
- Adds `cosmos_gov_version` column to ChainNodes table
- If a chain upgrades to gov v1, this automatically upgrades our support for them.

## "How We Fixed It"

- When a default v1beta1 proposal fetch provides a specific v1 response, flags this chainNode as "v1beta1-attempt-failed"
- Upon refresh, attempts `v1` fetch. If successful, updates `cosmos_gov_version` to "v1"

## Test Plan
- Added playwright tests for this case (and fixed existing ones that were broken)
  - these do not run in CI. To run locally, add command to package.json:
  - `"test-e2e-stateful": "TEST_ENV=playwright npx playwright test --workers 1 ./test/e2e/e2eStateful/*",`
  - run `yarn db-all` if your DB was deleted by other test runs (these stateful tests need db entities)
  - run `test-e2e-stateful` and expect proposals.spec to pass
- CA (click around) tested on local (steps are the same as in playwright tests):
  - in local DB, run migration, then to test this run:
    - `UPDATE "ChainNodes" set "cosmos_gov_version" = null where cosmos_chain_id = 'qwoyn';`
    - `UPDATE "ChainNodes" set "alt_wallet_url" = null where cosmos_chain_id = 'qwoyn';`
  - go to http://localhost:8080/qwoyn-network/proposals - expect proposal load to fail
  - refresh - expect proposals to load properly
  - refresh again - expect the same
  - Optional: check db to confirm changes for qwoyn: Community.`cosmos_gov_version` should be "v1"

## Deployment Plan

1. Keep existing `COSMOS_GOV_V1` env var for now. The new Notifications system still references it.

## Other Considerations

- Future consideration: We may want to clean up other code that relies on `COSMOS_GOV_V1` env var. Now using DB to keep track of gov version.
- The effect of this PR is that if any v1 community initially gets an error because of using the v1beta1 API, it will automatically use the v1 API in the next request. So only one person should ever see the proposal load fail. It will succeed the next time.
- However, this just catches one specific known error. There are also cases where backwards-compatibility still works for v1 chains. If a community is in this case, they may not see any problems. But if they prefer to have v1 support, we can upgrade them manually with a simple DB update: Community `cosmos_gov_version` to "v1"
